Improve a11y of the locale chooser

tdeschryver opened the pull request:
**What**: see #184 

**Why**: To make it accessible

**How**: 
- Highlighting: re-added outline on button, changed background to the `link` instead of the `list item`
- `Aria` tags: added `aria-labels` to `Links` and also `alt` to flag icons
- Close on click outside: used `componentDidMount` and `componentWillUnmount` to bind a `click` event to the document. Maybe there are some improvements here.

**Note**:
Also changed `getLocaleAndHost` to `getHost`, because the locale wasn't needed anymore (was a left over from #177)


![a11y](https://user-images.githubusercontent.com/28659384/27760301-615a2048-5e44-11e7-9688-a2195b4c746c.gif)

lydell commented on components/locale-chooser.js:
```diff
@@ -85,16 +88,31 @@ class LocaleChooser extends React.Component {
+  }
+
+  componentWillUnmount() {
+    document.removeEventListener('click', this.clickOutside.bind(this), true)
```
`this.clickOutside.bind(this)` returns an entirely new function, so this won’t actually remove the event listener added in `componentDidMount`.

lydell commented on components/locale-chooser.js:
```diff
@@ -85,16 +88,31 @@ class LocaleChooser extends React.Component {
 
-  handleClick() {
+  componentDidMount() {
+    document.addEventListener('click', this.clickOutside.bind(this), true)
```
This is a bit misleading since `clickOutside` is fired even when clicking _inside_ the dropdown.

Anujan commented on components/__tests__/__snapshots__/locale-chooser.js.snap:
```diff
@@ -334,21 +329,25 @@ exports[`opened state 1`] = `
                   >
                     <div>
                       <EsSvg
+                        alt="Español"
```
It would be beneficial to add `lang="es"` where you use languages that don't match the `html[lang]` attribute so the screen reader knows what language to interpret it as.

karlgroves replied:
+1 Add `lang="es"` on the SVG

tdeschryver replied:
Is this still needed after adding `aria-hidden="true"` on the svg?

Anujan commented on components/__tests__/__snapshots__/locale-chooser.js.snap:
```diff
@@ -274,21 +265,25 @@ exports[`opened state 1`] = `
+                  aria-label="Change localization to English"
+                  href="about://undefinedblank"
                 >
                   <a
```
Does this go `a[aria-label="Change localization to English"] > div > svg[alt="English"]`? If so, I'd recommend changing the the `alt` attribute to be blank on the SVGs here so they're considered decorative.

Belelros replied:
Also add `role="presentation"` for better interoperability as stated [here](https://www.w3.org/WAI/tutorials/images/decorative/).

karlgroves replied:
The suggestion of adding `aria-hidden="true"` on the SVG eliminates the need for `role="presentation"`

Anujan commented on components/locale-chooser.js:
```diff
@@ -138,19 +155,19 @@ function mapLocale(key) {
         key,
         display: 'English',
-        flag: <EnSvg {...svgStyle} />,
+        Flag: <EnSvg alt="English" {...svgStyle} />,
```
Here too. alt attribute should be blank since the SVGs are decorative

kentcdodds commented:
Thanks for the review and comments @Anujan!

Belelros commented:
This is looking great. A few minor comments that I think could improve this even further.

Perhaps I got this wrong since the button seems a bit funky. However I'd add a label to the button that opens the selector with something like "Change language localization" or something similar.

Upon opening the selector, focus either the list of the first element (with focus) because the focus remains within the button.

Span within the list element should probably have `aria-hidden="true"` so it's not read.  The anchor already has the correct label: `Change localization to English` reading `English` would be redundant.

<img width="548" alt="screenshot 2017-07-01 16 08 34" src="https://user-images.githubusercontent.com/946645/27762733-e2dd675a-5e77-11e7-8a12-391a6f723e75.png">

Close with ESC would be a nice touch. Return focus to button after that.

kentcdodds commented:
Great feedback! Much appreciated!

If you notice anything else on the site, feedback would likewise be appreciated!

https://rc.glamorous.rocks

lydell commented:
Navigating through the menu using the keyboard I kinda expected the menu to close when pressing Escape. ¯\\\_(ツ)\_/¯

karlgroves commented on components/__tests__/__snapshots__/locale-chooser.js.snap:
```diff
@@ -334,21 +329,25 @@ exports[`opened state 1`] = `
                         width="1em"
                       >
                         <svg
+                          alt="Español"
```
Again, alt is not valid on SVG

karlgroves commented:
@kentcdodds can you put this somewhere for testing?

It appears that you're intending this to behave as a menu. If that's the case, you'll want to adhere to 
https://www.w3.org/TR/wai-aria-practices-1.1/#menu

Pay close attention to the keyboard interaction patterns defined.  The version of this menu that is currently on the site does not adhere to the necessary patterns.

tdeschryver commented:
Thanks for the feedback everyone!
@karlgroves Once I corrected all of the stuff, I'll try to spin up a testing environment with [now](https://zeit.co/now).

tdeschryver commented:
I think I changed most of the remarks. Thanks for the feedback!
- removed `alt` tags on the flag `svg`'s
- added `aria-label` on the button
- added `lang` on `anchor`
- added `aria-hidden` on div around `svg` and display `span`, is this OK? (comments were that the SVG flag needed to have it)
- escape closes the locale picker and focuses the button
- refactored the click outside
- when opening the locale picker, the first item is being opened. Can this be reviewed please, I think there must be a better way of doing it...

@karlgroves When trying to set up the test environment the build process kept spitting out errors. I don't know why at the moment, but I'll take a look at it tomorrow if its still needed.

kentcdodds commented:
I added a script for building and deploying. See the CONTRIBUTING.md for instructions 👍

tdeschryver commented:
@kentcdodds I'm not able to build.
First had an issue with `integrations/index` (`cannot read property 'type' of undefined @ babel-generator\lib\whispace.js:50:19`).
And now I'm getting the error `Error: EEXIST, file already exists C:\...\glamorous-website\out\en\_next\0d3594bb-ff23-4d69-8c1a-9f70af64ff30\page\_document\index.js
    at fsError (C:\...\glamorous-website\node_modules\recursive-copy\lib\copy.js:377:15)
    at C:\...\glamorous-website\node_modules\recursive-copy\lib\copy.js:252:16
    at <anonymous>`. 

Tried reinstalling everything in a clean directory but without any success, any thoughts?

kentcdodds commented:
Hmmm... That's pretty weird. Ummm.... Yeah, I'm not sure what to try, sorry 😕

I'll try to pull your branch and deploy it myself later.

kentcdodds commented:
Here it is: https://en-glamorous-hhttuorgyn.now.sh

I've rebased on master as well...

I think that choosing a language and redirecting is broken.. 🤔 

Thanks!

karlgroves commented:
This will still need some work.

Here is the markup for a menu on the Tenon.io website:

```<!--// This LI container is part of the overall nav //-->
<li classname="dropdown">

	<button aria-haspopup="true" aria-owns="myAccountMenu" aria-label="Open Account menu" aria-expanded="true">
		<!--// Snipped other stuff, for clarity //-->
	</button>

	<ul id="myAccountMenu" role="menu" aria-label="Account Menu" tabindex="-1">

		<li role="menuitem" tabindex="-1">
			<a href="/dashboard.php#/project">Dashboard</a>
		</li>

		<li role="menuitem" tabindex="-1">
			<a href="/admin/">Admin</a>
		</li>

		<li role="menuitem" tabindex="-1">
			<a href="/editprofile.php">Profile</a>
		</li>

		<li role="menuitem" tabindex="-1">
			<a href="/messages.php">Notifications</a>
		</li>

		<li role="menuitem" tabindex="-1">
			<a href="/logout.php">Logout</a>
		</li>

	</ul>

</li>```

Note the menu/ menuitem roles, etc.

Also, take special care to read the Keyboard Interaction section of this:
https://www.w3.org/TR/wai-aria-practices-1.1/#menu

tdeschryver commented:
@kentcdodds Do I take the localization of the menu bar (with the input of @karlgroves) in this PR?

tdeschryver commented:
Is there a reason why the locale picker is being shown at the top of the side menu?
Because on the homepage and on the mobile menu its being shown as the last item.
If it has to be shown at the top, when how does it have to behave when being opened (over the other menu items, other menu items need to be pushed to the bottom, ....)?

 I changed it locally to see how it would look like if it was the last menu item:
![image](https://user-images.githubusercontent.com/28659384/27773441-ff5875e8-5f79-11e7-8648-8d03635ab287.png)

kentcdodds commented:
At first, I thought it was better at the top, but now I think it's better at the bottom so I'm good with your changes

tdeschryver commented:
Going to push the changes to move it to the bottom.
Tested it quickly because I have to go. I'll test it thoroughly tomorrow.

tdeschryver commented:
I think this is ready for another review.
The menu recommendations from @karlgroves deserves an own PR in my opinion.
@kentcdodds For the snapshot testing I used `mount` instead of `render` because with the `render` method the list items in the locale selector weren't in the snapshot.

kentcdodds commented on components/locale-chooser.js:
```diff
@@ -51,68 +51,135 @@ const Item = glamorous.li((props, {colors, mediaQueries}) => ({
+          open={this.state.open}
+          top={this.props.top}
+        >
+          {this.state.locales.map(({key, display, Flag}) =>
```
Hmmm, should we have a condition to check for the `open` state?

tdeschryver replied:
To be honest I don't know.
I was thinking it would be better to hide the list with some `css`, because this way the DOM isn't exactly being re-rendered.
Based on your answer I was curious to see how others are doing it (jest and webpack), and they are doing it with `css`.

If you want to I can add a check on the `open` state.

kentcdodds replied:
> because this way the DOM isn't exactly being re-rendered.

Why are you concerned about the DOM being re-rendered? I guess I don't really care either way... But it makes the snapshot even worse. If we remove all the snapshots then I'm :+1: with just using CSS to hide it.

kentcdodds commented on components/__tests__/locale-chooser.js:
```diff
@@ -11,48 +11,111 @@ const {supportedLocales, fallbackLocale} = require('../../config.json')
+test('it renders', () => {
+  setHost()
+  const wrapper = mountLocaleChooser()
+  expect(toJson(wrapper)).toMatchSnapshotWithGlamor()
```
I think maintaining this snapshot will just be really difficult. I would rather this test be simply:

```
test('it renders', () => {
  setHost()
  mountLocaleChooser()
})
```

Or just remove it entirely.

tdeschryver replied:
Removed it.

kentcdodds commented on components/locale-chooser.js:
```diff
@@ -121,48 +188,57 @@ function localeToHref(locale) {
+  return host
 }
 
-function mapLocale(key = fallbackLocale) {
```
I really dislike switch statements. They're overly verbose. Could we go back to the object map?

tdeschryver replied:
We sure can! 
Sorry, Must have overlooked it while rebasing.

kentcdodds commented on components/locale-chooser.js:
```diff
@@ -121,48 +188,57 @@ function localeToHref(locale) {
+      return {
+        key,
+        display: 'English',
+        Flag: <EnSvg {...options} />,
```
Could we leave most of this stuff the way it was? I liked the way the rendering of `Flag` better the way it was before.

kentcdodds commented on components/nav.js:
```diff
@@ -130,11 +130,6 @@ class Nav extends React.Component {
               {content.home}
             </Hidden>
           </ListItemAnchor>
-          {top ?
```
This change is :+1:
